django/db/backends/base/schema.py

```diff
@@ -268,6 +269,14 @@ def table_sql(self, model):
             constraint.constraint_sql(model, self)
             for constraint in model._meta.constraints
         ]
+
+        # If the model has a composite primary key, add the PRIMARY KEY constraint
+        # to the table definition.
+        # It's expected 'primary_key=True' isn't set on any fields (see E043).
+        pk = model._meta.pk
+        if hasattr(pk, "columns"):
```

Comment:
This might have been answered somewhere else already, so please ignore the question if it was. I am wondering if it would make sense to handle single column primary keys like this as well. What would be the upside of having different objects (?) here for single vs multi-columns fields? Or do we gain some backwards compat wins here by having it like this?

Reply:
I don't think we've considered this yet. My feeling without digging in is that it's not necessary for this feature, but it might be a nice follow-up refactor.

Reply:
Ok, if we want to do it in a follow-up refactor we should still ensure that it ends up in the same release. While Meta is probably not documented, changes there will still hurt more advanced 3rd party packages.

Reply:
The only difference is, `primary_key=True` creates the pk inline, while `Meta.primary_key` creates it as a table level constraint.

```
create table foo as (
    Id integer primary key
)
```
vs.
```
create table foo as (
    id integer
    primary key (id)
)
```

Reply:
Afaik all db backends support both inline and table-level primary keys, so there's no real difference in practice. It would take some work to move away from inline primary keys and it's out of scope for this ticket.

Reply:
I am less concerned about the individual backend support and emitted SQL but about the python API which we should get rid with the first try if possible. We can simplify the SQL generation afterwards.

Reply:
Why do we need to get rid of inline primary keys though?

I suppose to achieve this we could introduce the `columns` property for all fields, and not create any inline primary keys anymore.

I could open a separate PR for this which as a pre-requisite for this PR if it's necessary.

Reply:
I'm now leaning towards not doing this. I think inline is likely to be more understandable to anyone looking at the db schema at a low level, so I don't mind some extra complexity in Django to use inline for single pks. Especially since that code already exists.

Reply:
Please note that I don't really care about the schema (yet), but I think it might make sense to unify the python API in the meta object  (`meta.pk`) to always expose an object with `columns` even if it is solely a single column. Especially if we add type hints at some points, a single API will be easier to type and requires less use of runtime inspection (`hasattr`). I haven't checked how much overhead this will result in for single column primary keys though and whether it makes code awkward somewhere else.

Reply:
Ah, yeah that still makes sense to investigate.

Reply:
I'm not sure I agree, this would mean we need to add a `columns` property to every single field type. I would find it confusing, it's not clear what purpose it would serve.

If `column` is set, Django creates a column for the field in the database. If `columns` is set, does that mean Django should create multiple fields in the database?

It's not how it's used here. Maybe checking `isinstance(pk, CompositePrimaryKey)` is more appropriate here?

Reply:
> I'm not sure I agree, this would mean we need to add a `columns` property to every single field type. I would find it confusing, it's not clear what purpose it would serve.

Not necessarily. We already have support for:

```python
class X(models.Model):
   p = models.IntegerField()

  class Meta:
     primary_key = ["p"]
```
 
What if we *always* convert (even when defined with `primary_key=True` on an individual field) to `primary_key = ["field", "list"]`. Then we would always have `.meta.pk` point to a `CompositePrimaryKey` and we could adjust accordingly.

And instead of naming the fields `_primary_key` on the model we could maybe even change `model.pk` (sic) to utilize that field. Then we would always have a `CompositePrimaryKey` as interface to the pk.

The `CompositePrimaryKey` should btw optimize in `get_lookup` to use the ordinary lookups if only a single field is passed.

> It's not how it's used here. Maybe checking `isinstance(pk, CompositePrimaryKey)` is more appropriate here?

That would certainly be a good start (also in models/base.py where you have `if hasattr(pk, "fields"):`), this would allow type checkers to do type narrowing as well as making it easier to find where the composite primary key is actually used. The current code feels rather fragile and requires at least a comments about the ideas behind the hasattr check.

So to summarize: From an enduser POV why is there so much difference between `primary_key=True` on a field named `test` and `primary_key=["test"]`. The docs even say (https://github.com/django/django/pull/18056/files#diff-cca8870fcaec19104d999f61553ba925c72e2eb19b4933068c4849f2ce58a6f6R35):

> This is practically the same as setting ``primary_key=True`` on the field.

So let's actually make it the same?

Reply:
I don't think we can really change `model.pk` or `meta.pk`, it would be a backwards-incompatible change

Reply:
Actually, I am second guessing if `Meta.primary_key` is really necessary now.

You do raise the entirely valid point, why is it `None` if `primary_key=True` is set on a field? As a user, I would expect it to always store some info related to the primary key. And I would expect this info to be the source of truth.

So, how about we consider the following API instead?

```
class Foo(models.Model):
   tenant_id = models.IntegerField()
   id = models.IntegerField()
   primary_key = models.CompositeField("tenant_id", "id", primary_key=True)
```

This is how @LilyFoote did it too in the previous PR, I just thought adding a Meta option is more user friendly.
But if it leads to confusion, I'm not sure it's worth it anymore.

Reply:
I adjusted it 👍

Reply:
Let me know what you think, and thanks for the review @apollo13 , I appreciate it! ❤️